File: src/path/src/octomap_checker.cpp
#include <ros/ros.h>
#include <octomap/octomap.h>
#include <octomap_msgs/Octomap.h>
#include <octomap_msgs/conversions.h>
#include <path/CheckCollision.h>
#include <mutex>

class OctomapChecker
{
public:
    OctomapChecker()
    {
        ros::NodeHandle nh;

        octree_ = nullptr;

        // Subscriber
        octomap_sub_ = nh.subscribe("octomap_binary", 1, &OctomapChecker::octomapCallback, this);

        // Service
        service_ = nh.advertiseService("check_collision_at_point", &OctomapChecker::handleCollisionCheck, this);

        ROS_INFO("Octomap checker service is ready.");
    }

private:
    std::shared_ptr<octomap::OcTree> octree_;
    ros::Subscriber octomap_sub_;
    ros::ServiceServer service_;
    std::mutex mtx_;

    void octomapCallback(const octomap_msgs::Octomap::ConstPtr& msg)
    {
        std::lock_guard<std::mutex> lock(mtx_);

        try {
            octomap::AbstractOcTree* tree = octomap_msgs::binaryMsgToMap(*msg);
            if (tree) {
                // dynamic_cast to OcTree
                octomap::OcTree* octree = dynamic_cast<octomap::OcTree*>(tree);
                if (octree) {
                    octree_.reset(octree);
                    ROS_INFO_ONCE("Successfully received and parsed first OctoMap message.");
                } else {
                    ROS_ERROR("Received Octomap is not an OcTree!");
                    delete tree;
                }
            }
        } catch (std::exception& e) {
            ROS_ERROR("Failed to parse Octomap message: %s", e.what());
        }
    }

    bool handleCollisionCheck(path::CheckCollision::Request& req,
                              path::CheckCollision::Response& res)
    {
        std::lock_guard<std::mutex> lock(mtx_);

        if (!octree_) {
            ROS_WARN_THROTTLE(5, "Octomap not yet received. Assuming collision for safety.");
            res.is_occupied = true;
            return true;
        }

        try {
            octomap::OcTreeNode* node = octree_->search(req.point.x, req.point.y, req.point.z);

            if (node && octree_->isNodeOccupied(node)) {
                res.is_occupied = true;
            } else {
                res.is_occupied = false;
            }

        } catch (std::exception& e) {
            ROS_ERROR("Error during Octomap search: %s", e.what());
            res.is_occupied = true;
        }

        return true;
    }
};

int main(int argc, char** argv)
{
    ros::init(argc, argv, "octomap_checker_node");

    OctomapChecker checker;

    ros::spin();
    return 0;
}

File: src/path/src/octomap_checker.py
import rospy
import octomap
import threading
from octomap_msgs.msg import Octomap
from capstone.srv import CheckCollision, CheckCollisionResponse

class OctomapChecker:
    def __init__(self):
        rospy.init_node('octomap_checker_node', anonymous=True)

        self.octrr = None
        self.lock = threading.Lock()

        self.octomap_sub = rospy.Subscriber('octomap_binary', Octomap, self.octomap_callback, queue_size=1)
        self.collision_service = rospy.Service('check_collision_at_point', CheckCollision, self.handle_collision_check)

        rospy.loginfo("Octomap checker service is ready.")

    def octomap_callback(self, msg):
        with self.lock:
            try:
                self.octree = octomap.OcTree.readBinary(msg.data)
                rospy.loginfo_once("Successfully received and parsed first OctoMap message.")
            except Exception as e:
                rospy.logerr("Failed to parse Octomap message: %s", e)
    
    def handle_collision_check(self, req):
        response = CheckCollisionResponse()

        with self.lock:
            if self.octrr is None:
                rospy.logwarn_throttle(5, "Octomap not yey received. Assuming collision for safety.")
                response.is_occupied = True
                return response
            
            point = (req.point.x, req.point.y, req.point.z)

            try:
                node = self.octree.search(point[0], point[1], point[2])

                if node and self.octree.isNodeOccupied(node):
                    response.is_occupied = True
                else:
                    response.is_occupied = False
            
            except Exception as e:
                rospy.logerr("Error during Octomap search: %s", e)
                
                response.is_occupied = True

        return response

if __name__ == '__main__':
    try :
        checker = OctomapChecker()
        rospy.spin()
    except rospy.ROSInterruptException:
        pass